File: solved/problem40.py

import pdb

def index(s, i):
	return s[i-1]

s = ""
for i in xrange(1,1000000):
	s += str(i)

prod = 1
l = [1,10,100,1000,10000,100000,1000000]
for e in l:
	prod *= int(index(s, e))
print prod

File: solved/problem40.hs

import qualified Data.Char as Char

s :: String
s = concatMap show [1..1000000]

index :: String -> Int -> Char
index s i = s !! (i - 1)

prod :: Int
prod
	= product
	. map (Char.digitToInt . index s)
	$ [1,10,100,1000,10000,100000,1000000]

main :: IO ()
main = do
	putStrLn . show $ prod